### ISSUE TITLE
CDI device injection failed: unresolvable CDI devices

### ISSUE THREAD

**Issue by cyclinder** [moby/moby#47633]
### Description

can't run a container with cdi enabled

### Reproduce

```
root@worker-node-1:/home/cyclinder# docker run -it --network=host -d --device spidernet.io/test=demo1 docker.io/library/ubuntu:22.04   bash
1af7bceeae09fe86271388dc13fb5b3bee8dd9ff656915cb80c81d75ac178b4c
docker: Error response from daemon: CDI device injection failed: unresolvable CDI devices spidernet.io/test=demo1.
root@worker-node-1:/home/cyclinder# cat /var/run/cdi/test.yaml
cdiVersion: 0.6.0
kind: spidernet.io/test
containerEdits:
  env:
  - LD="test"
devices:
  - name: demo1
```

### Expected behavior

able to run a container with cdi enabled

### docker version

```bash
root@worker-node-1:/home/cyclinder# docker version
Client: Docker Engine - Community
 Version:           26.0.0
 API version:       1.45
 Go version:        go1.21.8
 Git commit:        2ae903e
 Built:             Wed Mar 20 15:17:48 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          26.0.0
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.8
  Git commit:       8b79278
  Built:            Wed Mar 20 15:17:48 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    26.0.0
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.13.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.25.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 10
  Running: 0
  Paused: 0
  Stopped: 10
 Images: 2
 Server Version: 26.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 CDI spec directories:
  /etc/cdi
  /var/run/cdi
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-94-generic
 Operating System: Ubuntu 22.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 56
 Total Memory: 125.8GiB
 Name: worker-node-1
 ID: e5f56c6c-2716-4776-9f98-23bbcc165bab
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
This error comes from the CDI dependency that's used; https://github.com/moby/moby/blob/bfdb8918f93f021c1ef4262b909a090518e8a637/vendor/tags.cncf.io/container-device-interface/pkg/cdi/cache.go#L227-L248

Wondering though; comparing your example with the example from https://github.com/moby/moby/pull/45134, should the `containerEdits` be part of the `demo1` device? i.e., more like;


```yaml
cdiVersion: 0.6.0
kind: spidernet.io/test
devices:
  - name: demo1
    containerEdits:
      env:
        - LD="test"
```


I tried reproduce your issue in a docker-in-docker container;


Creating the CDI directories, and `/etc/docker` for the daemon config file;

```bash
mkdir -p /etc/cdi /var/run/cdi /etc/docker
```

Configure the daemon;

```bash
cat > /etc/docker/daemon.json <<'EOF'
{
  "features": {
     "cdi": true
  }
}
EOF
```

Create the CDI spec based on your example;

```bash
cat > /var/run/cdi/test.yaml <<'EOF'
cdiVersion: 0.6.0
kind: spidernet.io/test
containerEdits:
  env:
  - LD="test"
devices:
  - name: demo1
EOF
```

Start the daemon;

```bash
dockerd
```

In another shell; run a container;

```bash
docker run -it --rm --device spidernet.io/test=demo1 alpine printenv LD
```

Check the daemon logs, and I see it complains that the spec is invalid;

```
WARN[2024-03-26T11:39:44.498974429Z] Refreshing the CDI registry generated errors  error="1 error occurred:\n\t* failed to load CDI Spec invalid CDI Spec: failed add device \"demo1\": invalid device, empty device edits\n\n"
```


That error looks to confirm my suspicion that your CDI spec is incorrect; changing the spec to what I _think_ it should be (see my comment further up);

```bash
cat > /var/run/cdi/test.yaml <<'EOF'
cdiVersion: 0.6.0
kind: spidernet.io/test
devices:
  - name: demo1
    containerEdits:
      env:
        - LD="test"
EOF
```

This time it worked, and the `LD` env-var is set to `"test"`;

```bash
docker run -it --rm --device spidernet.io/test=demo1 alpine printenv LD
"test"
```

**Comment by thaJeztah** [comment]
Based on the above, I don't think there's a bug at hand here, so I'll go ahead and close this ticket, but feel free to continue the conversation.

**Comment by cyclinder** [comment]
Thanks @thaJeztah for the details! This should be a bias in my use, perhaps I should go to the CDI community for help.

**Comment by thaJeztah** [comment]
Thanks! Out of curiosity; was your example `test.yaml` based on an example you found somewhere in documentation (so the documentation being incorrect?) or was this something you wrote yourself?

**Comment by cyclinder** [comment]
> was this something you wrote yourself?

woo, I wrote this myself.  when I read the CDI docs on https://github.com/cncf-tags/container-device-interface/blob/main/SPEC.md#oci-edits, I found the containerEdits field is referenced in two places in the specification:

- At the device level, where the edits MUST only be made if the matching device is requested by the container runtime user.
- At the container level, where the edits MUST be made if any of the device defined in the devices field are requested.

So even if I don't have containerEdits configured in devices, but I do have container-level containerEdits configured, the runc will be able to take effect on the container-level containerEdits,  But it's not working. maybe I misunderstood something.
